core-api/src/main/java/com/optimizely/ab/event/BatchEventProcessor.java

```diff
@@ -0,0 +1,232 @@
+                thread.setDaemon(true);
+                return thread;
+            });
+        } else {
+            this.executor = executor;
+        }
+
+        start();
+    }
+
+    public synchronized void start() {
+        if (isStarted) {
+            logger.info("Executor already started.");
+            return;
+        }
+
+        isStarted = true;
+        EventConsumer runnable = new EventConsumer();
+        future = executor.submit(runnable);
+    }
+
+    @Override
+    public void close() throws Exception {
+        logger.info("Start close");
+        eventQueue.put(SHUTDOWN_SIGNAL);
+        try {
+            future.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            logger.warn("Interrupted while awaiting termination.");
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    public void process(UserEvent userEvent) {
+        logger.debug("Received userEvent: {}", userEvent);
+
+        if (executor.isShutdown()) {
+            logger.warn("Executor shutdown, not accepting tasks.");
+            return;
+        }
+
+        if (!eventQueue.offer(userEvent)) {
+            logger.warn("Payload not accepted by the queue. Current size: {}", eventQueue.size());
+        }
+    }
+
+    public class EventConsumer implements Runnable {
+        private LinkedList<UserEvent> currentBatch = new LinkedList<>();
+        private long deadline = System.currentTimeMillis() + flushInterval;
+
+        @Override
+        public void run() {
+            try {
+                while (true) {
+                    if (System.currentTimeMillis() > deadline) {
+                        logger.debug("Deadline exceeded flushing current batch.");
+                        flush();
+                    }
+
+                    Object item = eventQueue.poll(50, TimeUnit.MILLISECONDS);
+                    if (item == null) {
+                        logger.info("Empty item, sleeping for 50ms.");
+                        Thread.sleep(50);
+                        continue;
+                    }
+
+                    if (item == SHUTDOWN_SIGNAL) {
+                        logger.info("Received shutdown signal.");
+                        break;
+                    }
+
+                    addToBatch((UserEvent) item);
+                }
+            } catch (InterruptedException e) {
+                logger.info("Interrupted while processing buffer.");
+            } catch (Exception e) {
+                logger.error("Uncaught exception processing buffer.", e);
+            } finally {
+                logger.info("Exiting processing loop. Attempting to flush pending events.");
+                flush();
+            }
+        }
+
+        private void addToBatch(UserEvent userEvent) {
+            if (!currentBatch.isEmpty()) {
+                String currentRevision = currentBatch.peekLast().getUserContext().getProjectConfig().getRevision();
```

Comment:
One observation regarding this. The current set up allows me to create 1 instance of `BatchEventProcessor` and use it in multiple Optimizely instances each of which may be tied to different projects. It is certainly possible that a customer does such a thing.

In that case, I was thinking if:
- we want to somehow disallow that? or
- flush based on both revision and project ID.